python/seldon_core/microservice.py

```diff
@@ -351,107 +333,108 @@ def main():
     else:
         user_object = user_class(**parameters)
 
-    port = args.port
+    http_port = args.http_port
+    grpc_port = args.grpc_port
     metrics_port = args.metrics_port
 
-    if args.tracing:
-        tracer = setup_tracing(args.interface_name)
+    # if args.tracing:
+    #    tracer = setup_tracing(args.interface_name)
+
+    seldon_metrics = SeldonMetrics(worker_id_func=os.getpid)
+    # TODO why 2 ways to create metrics server
+    # seldon_metrics = SeldonMetrics(
+    #    worker_id_func=lambda: threading.current_thread().name
+    # )
+    if args.debug:
+        # Start Flask debug server
+        def rest_prediction_server():
+            app = seldon_microservice.get_rest_microservice(user_object, seldon_metrics)
+            try:
+                user_object.load()
+            except (NotImplementedError, AttributeError):
+                pass
+            if args.tracing:
+                logger.info("Tracing branch is active")
+                from flask_opentracing import FlaskTracing
 
-    if args.api_type == "REST":
-        seldon_metrics = SeldonMetrics(worker_id_func=os.getpid)
+                tracer = setup_tracing(args.interface_name)
 
-        if args.debug:
-            # Start Flask debug server
-            def rest_prediction_server():
-                app = seldon_microservice.get_rest_microservice(
-                    user_object, seldon_metrics
-                )
-                try:
-                    user_object.load()
-                except (NotImplementedError, AttributeError):
-                    pass
-                if args.tracing:
-                    logger.info("Tracing branch is active")
-                    from flask_opentracing import FlaskTracing
-
-                    logger.info("Set JAEGER_EXTRA_TAGS %s", jaeger_extra_tags)
-                    FlaskTracing(tracer, True, app, jaeger_extra_tags)
-
-                app.run(
-                    host="0.0.0.0",
-                    port=port,
-                    threaded=False if args.single_threaded else True,
-                )
+                logger.info("Set JAEGER_EXTRA_TAGS %s", jaeger_extra_tags)
+                FlaskTracing(tracer, True, app, jaeger_extra_tags)
 
-            logger.info(
-                "REST microservice running on port %i single-threaded=%s",
-                port,
-                args.single_threaded,
+            app.run(
+                host="0.0.0.0",
+                port=http_port,
+                threaded=False if args.single_threaded else True,
```

Comment:
In the case when users may want to run single-threaded single-process due to non-thread safe models like pytorch-based ones, would it still work even if always both the grpc and rest servvers would be spinned up?

Reply:
hmm... they would both be spun up but fairly independent so I guess it should still work, here for REST server and bellow would be set for gRPC